security/advancedtls/crl.go

```diff
@@ -83,15 +88,48 @@ func (s RevocationStatus) String() string {
 	return [...]string{"RevocationUndetermined", "RevocationUnrevoked", "RevocationRevoked"}[s]
 }
 
-// certificateListExt contains a pkix.CertificateList and parsed
+// CRL contains a pkix.CertificateList and parsed
 // extensions that aren't provided by the golang CRL parser.
-type certificateListExt struct {
+// GRPC requires certain specifics for CRLs - all CRLs should be loaded using
+// NewCRL() for bytes directly or ReadCRLFile() to read directly from a filepath
+type CRL struct {
 	CertList *x509.RevocationList
 	// RFC5280, 5.2.1, all conforming CRLs must have a AKID with the ID method.
 	AuthorityKeyID []byte
 	RawIssuer      []byte
 }
 
+// NewCRL constructs new CRL from provided byte array.
+func NewCRL(b []byte) (*CRL, error) {
+	crl, err := parseRevocationList(b)
+	if err != nil {
+		return nil, fmt.Errorf("parseCrl() failed err = %v", err)
+	}
+	crlExt, err := parseCRLExtensions(crl)
+	if err != nil {
+		return nil, fmt.Errorf("parseCRLExtensions() failed err = %v", err)
+	}
+	crlExt.RawIssuer, err = extractCRLIssuer(b)
+	if err != nil {
+		return nil, fmt.Errorf("extractCRLIssuer() failed err= %v", err)
+	}
+	return crlExt, nil
+}
+
+// ReadCRLFile reads a file from the provided path, and returns constructed
```

Comment:
suggest something like "ReadCRLFile reads a crl file from the provided path and returns a constructed CRL struct."

Reply:
Done